File: apps/agent/src/agent/tools/list-commitments.ts
import { getCollections } from "../../lib/mongo.js";
import { listCommitmentRecords, publicCommitment } from "../../lib/commitments.js";
import type { ToolDef } from "../types.js";

export const listCommitmentsTool: ToolDef = {
  declaration: {
    name: "list_commitments",
    description:
      "List open commitments — promises Alex made (outgoing) or that others made to Alex (incoming). Use this for the commitment ledger, or when a question asks 'what do I owe X' / 'who owes me'.",
    parameters: {
      type: "OBJECT",
      properties: {
        direction: {
          type: "string",
          enum: ["incoming", "outgoing", "all"],
          description: "Filter to commitments owed TO Alex (incoming), owed BY Alex (outgoing), or all.",
        },
        actor: {
          type: "string",
          description: "Optional case-insensitive substring match on the other party's name.",
        },
        limit: {
          type: "integer",
          description: "Max commitments to return. Default 12, max 50.",
        },
      },
    },
  },
  handler: async (args) => {
    try {
      const direction =
        typeof args["direction"] === "string"
          ? (args["direction"] as "incoming" | "outgoing" | "all")
          : "all";
      const actor = typeof args["actor"] === "string" ? args["actor"].trim() : "";
      const limit = clampInt(args["limit"], 12, 1, 50);

      // ── Primary path: the persisted, LLM-extracted ledger. ──
      const records = await listCommitmentRecords({
        direction,
        ...(actor ? { actor } : {}),
        status: "open",
        limit,
      });
      if (records.length > 0) {
        return {
          ok: true,
          data: {
            direction,
            actor: actor || null,
            count: records.length,
            source: "ledger",
            commitments: records.map(publicCommitment),
          },
          summary: `${records.length} ${direction === "all" ? "" : direction + " "}commitments${actor ? ` involving ${actor}` : ""} · ledger`,
        };
      }

      // ── Fallback: regex heuristic over chunk text (used when the ledger
      // hasn't been built yet, so the demo still shows something). ──
      const { chunks } = await getCollections();
      const commitmentRegex =
        "(owe[ds]?|owed|will deliver|by Friday|by Mon|by Tue|by Wed|by Thu|by EOD|by next|committed to|promised|action item|owner:|due\\s)";
      const filter: Record<string, unknown> = {
        text: { $regex: commitmentRegex, $options: "i" },
      };
      if (actor) {
        filter["text"] = {
          $regex: `${commitmentRegex}.*${escapeRegex(actor)}|${escapeRegex(actor)}.*${commitmentRegex}`,
          $options: "is",
        };
      }

      const raw = await chunks
        .find(filter, {
          projection: { _id: 1, title: 1, source: 1, text: 1, metadata: 1 },
          limit: limit * 3,
        })
        .toArray();

      const inferred = raw.map((c) => ({
        chunkId: String(c._id),
        title: c.title,
        source: c.source,
        excerpt: typeof c.text === "string" ? trimToCommitment(c.text) : null,
        date: (c.metadata?.["date"] as string | undefined) ?? null,
        thread: (c.metadata?.["threadId"] as string | undefined) ?? null,
        direction: classifyDirection(typeof c.text === "string" ? c.text : ""),
      }));

      const filtered = inferred.filter((c) => direction === "all" || c.direction === direction).slice(0, limit);

      return {
        ok: true,
        data: {
          direction,
          actor: actor || null,
          count: filtered.length,
          commitments: filtered,
        },
        summary: `${filtered.length} ${direction === "all" ? "" : direction + " "}commitments${actor ? ` involving ${actor}` : ""}`,
      };
    } catch (err) {
      const msg = err instanceof Error ? err.message : String(err);
      return { ok: false, error: msg };
    }
  },
};

function classifyDirection(text: string): "incoming" | "outgoing" | "unknown" {
  const lower = text.toLowerCase();
  if (/\b(alex (will|to|owes|is delivering|to send|to ship|to deliver))\b/.test(lower)) {
    return "outgoing";
  }
  if (/\b(owes alex|to alex|will send alex|will deliver to alex|alex is waiting|alex expects)\b/.test(lower)) {
    return "incoming";
  }
  if (/\bi (will|owe|to)\b/.test(lower)) return "outgoing";
  return "unknown";
}

function trimToCommitment(text: string): string {
  const idx = text.search(
    /(owe|owed|will deliver|by Friday|by Mon|by Tue|by Wed|by Thu|by EOD|by next|committed|promised|action item|owner:|due\s)/i,
  );
  const start = idx >= 0 ? Math.max(0, idx - 60) : 0;
  return text.slice(start, start + 280).trim();
}

function escapeRegex(s: string): string {
  return s.replace(/[.*+?^${}()|[\]\\]/g, "\\$&");
}

function clampInt(raw: unknown, dflt: number, lo: number, hi: number): number {
  const n = typeof raw === "number" ? raw : Number(raw);
  if (!Number.isFinite(n)) return dflt;
  return Math.max(lo, Math.min(hi, Math.floor(n)));
}

File: apps/agent-py/app/agent/tools/list_commitments.py
"""list_commitments — port of apps/agent/src/agent/tools/list-commitments.ts.

Reads the persisted ledger first; falls back to a regex heuristic over chunk
text only when the ledger is empty.
"""
from __future__ import annotations

import re

from app.agent.types import ToolDef
from app.db.mongo import chunks
from app.lib.commitments import list_commitment_records, public_commitment

_DECL = {
    "name": "list_commitments",
    "description": (
        "List open commitments — promises Alex made (outgoing) or that others made to Alex (incoming). "
        "Use this for the commitment ledger, or when a question asks 'what do I owe X' / 'who owes me'."
    ),
    "parameters": {
        "type": "OBJECT",
        "properties": {
            "direction": {"type": "string", "enum": ["incoming", "outgoing", "all"],
                          "description": "Filter to commitments owed TO Alex, owed BY Alex, or all."},
            "actor": {"type": "string", "description": "Optional case-insensitive substring match on the other party."},
            "limit": {"type": "integer", "description": "Max commitments. Default 12, max 50."},
        },
    },
}

_RX = (r"(owe[ds]?|owed|will deliver|by Friday|by Mon|by Tue|by Wed|by Thu|by EOD|by next|"
       r"committed to|promised|action item|owner:|due\s)")


def _clamp(raw, default, lo, hi):
    try:
        return max(lo, min(hi, int(raw)))
    except (TypeError, ValueError):
        return default


def _classify(text: str) -> str:
    lower = text.lower()
    if re.search(r"\b(alex (will|to|owes|is delivering|to send|to ship|to deliver))\b", lower):
        return "outgoing"
    if re.search(r"\b(owes alex|to alex|will send alex|will deliver to alex|alex is waiting|alex expects)\b", lower):
        return "incoming"
    if re.search(r"\bi (will|owe|to)\b", lower):
        return "outgoing"
    return "unknown"


async def _handler(args: dict, ctx: dict | None = None) -> dict:
    try:
        direction = args["direction"] if isinstance(args.get("direction"), str) else "all"
        actor = args["actor"].strip() if isinstance(args.get("actor"), str) else ""
        limit = _clamp(args.get("limit"), 12, 1, 50)

        # Primary: persisted ledger
        records = await list_commitment_records(
            direction=direction, actor=actor or None, status="open", limit=limit)
        if records:
            return {"ok": True, "data": {
                "direction": direction, "actor": actor or None, "count": len(records),
                "source": "ledger", "commitments": [public_commitment(r) for r in records],
            }, "summary": f"{len(records)} {'' if direction == 'all' else direction + ' '}commitments"
                          f"{(' involving ' + actor) if actor else ''} · ledger"}

        # Fallback: regex over chunk text
        flt: dict = {"text": {"$regex": _RX, "$options": "i"}}
        if actor:
            esc = re.escape(actor)
            flt["text"] = {"$regex": f"{_RX}.*{esc}|{esc}.*{_RX}", "$options": "is"}
        rows = await chunks().find(
            flt, projection={"_id": 1, "title": 1, "source": 1, "text": 1, "metadata": 1}, limit=limit * 3
        ).to_list(length=None)
        inferred = []
        for c in rows:
            txt = c.get("text") if isinstance(c.get("text"), str) else ""
            md = c.get("metadata") or {}
            inferred.append({
                "chunkId": str(c["_id"]), "title": c.get("title"), "source": c.get("source"),
                "excerpt": txt[:280], "date": md.get("date"), "thread": md.get("threadId"),
                "direction": _classify(txt),
            })
        filtered = [c for c in inferred if direction == "all" or c["direction"] == direction][:limit]
        return {"ok": True, "data": {"direction": direction, "actor": actor or None,
                                     "count": len(filtered), "commitments": filtered},
                "summary": f"{len(filtered)} {'' if direction == 'all' else direction + ' '}commitments"}
    except Exception as err:  # noqa: BLE001
        return {"ok": False, "error": str(err)}


tool = ToolDef(declaration=_DECL, handler=_handler)
